### DISCUSSION TITLE
Dynamically call other commands from a command

### DISCUSSION THREAD

**Discussion by rsenden** [remkop/picocli#2249] [Q&A]
What's the best way to dynamically call other commands from a Picocli command implementation?

Suppose I have a sub-command named `run` that reads a YAML file containing instructions like the following:
```yaml
- cmd: subcmd1 subcmd2 --arg1 xyz
```

When encountering the `cmd` property above, I'd like to execute the given sub-command and retrieve it's output and exit code.

The following pseudocode seems to work for basic invocation (not explicitly handling command output yet), but not sure whether this is the best approach and whether this may cause any side-effects.

```java
@Command(name="run")
public class RunCommand implements Callable<Integer> {
  @Spec CommandSpec spec;

  public final Integer call() {
    return new YamlRunner(yamlFile, getRootCommandLine()).run();
  }

  private final CommandLine getRootCommandLine() {
    var cl = commandHelper.getCommandSpec().commandLine();
    while ( true ) {
      if ( cl.getParent()==null ) break;
        cl = cl.getParent();
    }
    return cl;
  }
} 

@RequiredArgsContructor
public class YamlRunner {
  private final String yamlFile;
  private final CommandLine rootCL;
  public int run() {
    ...
    return someExitCode;
  }
  private void handleCmd(cmdElt) {
    int exitCode = rootCL.execute(cmdElt.getArgs());
    // handle exitCode
  }
}
```

Ideally, I'd also like to have a generic mechanism to allow the sub-commands being run to pass data back to the `YamlRunner` through Java method calls. Is there any way to pass some Java object to the command being executed, for example using some execution context that can be retrieved through the `CommandSpec` (not sure whether something like this exists in picocli), or by calling an interface method on the command to be executed before and after execution?

**Comment by remkop** [comment]
If you want to pass Java objects via the memory of the Java runtime, you could set some static value and let each command read that static value for further processing. (The Java equivalent of a global variable.)

An alternative is to serialize the object to a file and pass the path to that file as a command line argument. 

That's all I can think of now.

**Reply by rsenden** [reply]
Thanks, I'm currently indeed using static methods/variables, I was just wondering whether there was any 'cleaner' approach, providing some kind of context that exists only during command execution.

Slightly cleaner might be to inject some value into the command being invoked, i.e., check whether the command to be invoked implements a certain interface, and if so, inject a value. Although I'm not injecting any value (yet), I do now use this approach to check what kind of command we're running, using code like the following:

```java
    public final CommandLine getRootCommandLine(CommandSpec spec) {
        var cl = spec.commandLine();
        while ( true ) {
            if ( cl.getParent()==null ) break;
            cl = cl.getParent();
        }
        return cl;
    }

    public final CommandSpec getLeafCommandSpec(ParseResult parseResult) {
        var leafCommand = parseResult.subcommand();
        while (leafCommand.hasSubcommand() ) {
            leafCommand = leafCommand.subcommand();
        }
        return leafCommand.commandSpec();
    }

    public final int execute(CommandLine rootCommandLine, ParseResult parseResult) {
        try {
            rootCommandLine.clearExecutionResults();
            return rootCommandLine.getExecutionStrategy().execute(parseResult);
        } catch ( Exception e ) {
            throw new SomeException("Command threw an exception", e);
        }
    }

// spec parameter comes from @Spec annotated field
// args comes from yaml file or some other source
var rootCommandLine = getRootCommandLine(spec)
var parseResult = rootCommandLine.parseArgs(args);
var leafCommandSpec = getLeafCommandSpec(parseResult);
var leafCommand = leafCommandSpec.userObject();
if ( leafCommand instanceof SomeInterface ( {
  // Do something, for example inject some object into leafCommand
}
var exitCode = execute(rootCommandLine, parseResult);
// Potential clean up, like removing/resetting objects injected into leafCommand
// Do something with exitCode
```

Of course, there's a lot more to it in the actual application, like intercepting stdout/stderr of the command being invoked, but the above is more or less all picocli-related code that I'm using to dynamically invoke commands from another command. 

Would this be the correct approach, or do you see any potential issues? Note that in this context, I'm not interested in generating help information if for example command options are incorrect.

**Comment by yuvkun10** [comment]
Your approach using `parseArgs` + `getExecutionStrategy().execute()` is actually the right pattern. To cleanly pass context objects to sub-commands without static variables, here are two better alternatives:

**Option 1: Use the `IFactory` to inject context into commands**

Picocli's `IFactory` is called to instantiate every command. You can provide a custom factory that injects your context:

```java
public class ContextAwareFactory implements IFactory {
    private final ExecutionContext context;

    public ContextAwareFactory(ExecutionContext context) {
        this.context = context;
    }

    @Override
    public <K> K create(Class<K> cls) throws Exception {
        try {
            // Try constructor injection first
            try {
                return cls.getDeclaredConstructor(ExecutionContext.class).newInstance(context);
            } catch (NoSuchMethodException e) {
                return CommandLine.defaultFactory().create(cls);
            }
        } catch (Exception e) {
            return CommandLine.defaultFactory().create(cls);
        }
    }
}

// Usage:
ExecutionContext ctx = new ExecutionContext();
CommandLine rootCL = new CommandLine(new RootCommand(), new ContextAwareFactory(ctx));
```

Sub-commands that need the context declare a constructor accepting it:

```java
@Command(name = "subcmd1")
public class SubCmd1 implements Callable<Integer> {
    private final ExecutionContext context;

    public SubCmd1(ExecutionContext context) {
        this.context = context;
    }

    public Integer call() {
        context.setResult("some output");
        return 0;
    }
}
```

**Option 2: Use `CommandLine.getUserObject()` after execution**

After executing a subcommand via `rootCL.execute(args)`, you can retrieve the command instance and call methods on it:

```java
private void handleCmd(String[] args) {
    ParseResult parseResult = rootCL.parseArgs(args);
    // Walk to leaf command
    ParseResult leaf = parseResult;
    while (leaf.hasSubcommand()) leaf = leaf.subcommand();

    Object commandObj = leaf.commandSpec().userObject();
    if (commandObj instanceof ResultCapture capture) {
        // Pre-execution setup
        capture.setExecutionContext(context);
    }
    rootCL.clearExecutionResults();
    int exitCode = rootCL.getExecutionStrategy().execute(parseResult);
    if (commandObj instanceof ResultCapture capture) {
        // Post-execution result retrieval
        Object result = capture.getResult();
    }
}
```

The `IFactory` approach (Option 1) is the cleanest because it uses picocli's built-in dependency injection mechanism and avoids any coupling to the `CommandSpec` internals.

Hope this helps! Let me know if you need clarification.

**Reply by rsenden** [reply]
@yuvkun10 Thanks for the suggestions! Given that I implemented this functionality in our code base about 2 years ago, I'd need to check whether utilizing `IFactory` will provide any benefits compared to the current implementation, but I'm sure this will come in handy for other use cases in our application.

**Reply by yuvkun10** [reply]
Glad it was helpful! Yeah, if your current approach is already working well in production for 2 years, there may not be a strong reason to refactor just for the sake of it. The `IFactory` approach mainly shines when you're setting up new commands or if you ever find yourself needing to pass different context objects to different commands — it keeps things decoupled without relying on static state.

Either way, your pattern of using `parseArgs` + `getExecutionStrategy().execute()` with the `instanceof` check for interface-based injection is solid. Feel free to reach out if you run into anything down the line!
